### PR TITLE
Implement `OperationStore` on test storage provider

### PR DESCRIPTION
**Pull request by sandreae**
This is required preparation for later implementing (and testing) storing `Operation`'s inside `publish_entry()`: https://github.com/p2panda/p2panda/issues/360 

- Implement `OperationStore` on the test storage provider `SimplestStorageProvider`
- use same `test_db` pattern as in `aquadoggo` [later this can be exported and made generic so it can be constructed with different backends]
- refactor tests to use new `test_db`

Follow-up: #366 

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header
- [x] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/operation/traits.rs [hunk 1]
```diff
@@ -61,7 +65,30 @@ pub trait AsOperation {
...
+    ///
+    /// Use `VerifiedOperation::new_from_entry()` instead if you want to validate that the operation
+    /// was signed by this public key.
+    fn new(
+        public_key: &Author,
+        operation_id: &OperationId,
+        operation: &Operation,
+    ) -> Result<Self, Self::VerifiedOperationError>;
+
+    /// Returns a new `VerifiedOperation` instance constructed from an `EntrySigned`
```
Comment (adzialocha) [inline p2panda-rs/src/operation/traits.rs:84]: This comment somehow suggests a functionality but it is not implemented as it is a trait. I would rephrase it as a suggestion and maybe someday we can even enforce some safety by how we design the traits.

I don't want to open a whole chapter here in this PR on that topic but I wonder how strict we should be around making validation promises and if our current api is reflecting that. It's maybe something to look at again together with the whole validation topic.

(There is a small typo in "oncoded").

File: p2panda-rs/src/storage_provider/traits/log_store.rs [hunk 2]
```diff
@@ -56,8 +56,14 @@ pub trait LogStore<StorageLog: AsStorageLog> {
 
 #[cfg(test)]
 pub mod tests {
+    // TEST MODULE NOTES //
+    //
+    // In this module we are implementing the `LogStore` trait on `SimplestStorageProvider` just
```
Comment (adzialocha) [inline p2panda-rs/src/storage_provider/traits/log_store.rs:61]: Im not sure if I understand this comment: aren't all tests written inside the `tests` module solely for testing purposes? 🤔

File: p2panda-rs/src/operation/traits.rs (hunk 1 above)
Comment (sandreae) [reply p2panda-rs/src/operation/traits.rs:84]: Yes, very good point.

In this particular case, these methods were only implemented on the struct before, and in this PR I moved them into the trait definition. So maybe the comments snuck across too :grimacing: .... I'll re-write the comment to be more trait-description-like. 

In general, yeh, this would be a very good conversation to have. I particularly get tripped up on where to place constructors. Sometimes they're needed in the trai (like here) but it feels a bit weird for some reason...

File: p2panda-rs/src/operation/traits.rs [hunk 3]
```diff
@@ -61,7 +65,30 @@ pub trait AsOperation {
...
-pub trait AsVerifiedOperation {
+pub trait AsVerifiedOperation:
+    Sized + Clone + Send + Sync + Validate + PartialEq + std::fmt::Debug
+{
+    /// Error type for `AsVerifiedOperation`
+    type VerifiedOperationError: 'static + std::error::Error + Send + Sync;
+
+    /// Returns a new `VerifiedOperation` instance.
+    ///
+    /// Use `VerifiedOperation::new_from_entry()` instead if you want to validate that the operation
```
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/operation/traits.rs:76]: Hmm, lolz, yeh this is weird. Is it just a naming thing? If this was called `AuthoredOperation` then it would be less weird? 

Maybe the method names could be reversed? `new()` takes entry and operation and verifies. Maybe we can place the other behind a testing flag actually, I don't think it is hugely useful in the wild. These methods aren't actually new in this PR though. So maybe I should open an issue and deal with this in a separate PR.

File: p2panda-rs/src/storage_provider/traits/log_store.rs (hunk 2 above)
Comment (sandreae) [reply p2panda-rs/src/storage_provider/traits/log_store.rs:61]: Yeh, it clear without the comment why the traits are being implemented here in the test module? It's for us to test the methods with default implementations, and also so we can test the higher level methods on `StorageProvider`? This is what I was trying (obviously not very well) to explain.

File: p2panda-rs/src/operation/traits.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/operation/traits.rs:76]: I raised an issue for this: https://github.com/p2panda/p2panda/issues/366

File: p2panda-rs/src/storage_provider/traits/log_store.rs (hunk 2 above)
Comment (adzialocha) [reply p2panda-rs/src/storage_provider/traits/log_store.rs:61]: Traits in a library crate are weird indeed. You provide an interface to the outside you don't use yourself (inside the library), except of maybe tests and partly in our some default implementations. I think its safe to assume that this is "normal" when testing traits in a library.

File: p2panda-rs/src/operation/traits.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandreae) [reply p2panda-rs/src/operation/traits.rs]: I'm gunna deal with this in a follow-up PR as the issue wasn't introduced in this PR.

File: p2panda-rs/src/storage_provider/traits/operation_store.rs [hunk 4]
```diff
@@ -49,5 +49,247 @@ pub trait OperationStore<Operation: AsVerifiedOperation> {
...
+            .await
+            .unwrap()
+            .is_none());
+
+        db.store
+            .insert_operation(&create_operation, &document_id)
+            .await
+            .unwrap();
+
+        assert_eq!(
```
Comment (cafca) [inline p2panda-rs/src/storage_provider/traits/operation_store.rs:237]: Wondering why this line is reported as not being hit 🤔

File: p2panda-rs/src/operation/verified_operation.rs [hunk 5]
```diff
@@ -27,12 +27,14 @@ pub struct VerifiedOperation {
...
-impl VerifiedOperation {
+impl AsVerifiedOperation for VerifiedOperation {
+    type VerifiedOperationError = VerifiedOperationError;
+
     /// Returns a new `VerifiedOperation` instance.
     ///
     /// Use `VerifiedOperation::new_from_entry()` instead if you want to validate that the operation
     /// was signed by this public key.
-    pub fn new(
+    fn new(
```
Comment (cafca) [inline p2panda-rs/src/operation/verified_operation.rs:37]: I'm still a bit unsure about documenting traits vs documenting structs but maybe trait items like this should have a description of what implementations need to verify - or don't need to verify. In this case it could say something like:

> Implementations should document that this method doesn't validate whether it's parameters match each other and that consumers must ensure that parameters were validated before using this constructor.

**Review by cafca — APPROVED** [review]
I had one suggestion for a trait item docstring but it's good!

File: p2panda-rs/src/operation/verified_operation.rs (hunk 5 above)
Comment (sandreae) [reply p2panda-rs/src/operation/verified_operation.rs:37]: Yes, I like that!

Comment (sandreae) [reply p2panda-rs/src/operation/verified_operation.rs:37]: I assumed you meant changing the docs in the trait definition?

Comment (sandreae) [reply p2panda-rs/src/operation/verified_operation.rs:37]: Hmm, also, thinking again. Not validating here isn't _compulsory_, if an implementation can do some validation somehow, this is fine. The thing we want to strongly encourage is that `new_from_entry)` implementations _do_ perform validation.

File: p2panda-rs/src/storage_provider/traits/operation_store.rs (hunk 4 above)
Comment (sandreae) [reply p2panda-rs/src/storage_provider/traits/operation_store.rs:237]: Yeh, I don't know either... I've seen this behaviour before too...

File: p2panda-rs/src/operation/verified_operation.rs (hunk 5 above)
Comment (sandreae) [reply p2panda-rs/src/operation/verified_operation.rs:37]: What do you think @cafca ? Is this good to go as is?

Comment (cafca) [reply p2panda-rs/src/operation/verified_operation.rs:37]: Yes!
